### ISSUE TITLE
Clip gradient before optimizer update

### ISSUE THREAD

**Issue by ringgaard** [google/sling#237]
The gradient clipping in learning.cc should be applied to the incoming gradient (dv) instead of the final update to be compatible with TF and PyTorch. See #235.
